### ISSUE TITLE
finish subviews/* work

### ISSUE THREAD

**Issue by lynxlynxlynx** [gemrb/gemrb#95]
The [branch](https://github.com/gemrb/gemrb/tree/subviews) accumulated tons of GUI, but also some unrelated work. It includes a windowing rewrite that also upgraded text handling and drawing significantly.

See the [branches bugtracker](https://github.com/bradallred/gemrb/issues) (started before the main repo had this one) and the [subviews label](https://github.com/gemrb/gemrb/labels/subviews) here. We could also [move the old ones](https://help.github.com/en/articles/transferring-an-issue-to-another-repository) over and relabel them.

Stuff to reverify (merge problems):
- n/a

Stuff to do after the merge:
- [x] change the github actions coverity workflow to check master again
- [x] remove subviews bug reporting template
- [x] recheck (closed) subviews-recheck bugs, since many will automatically close
- [x] archive subviews tag to the end of the list
- [x] add big demo data fetching to builder workflows
- [x] update NEWS

Among other things, this should fix the following reports (recheck when done):
- [x] desensitivize portrait dragging
- [x] portrait icons flicker (check in iwd1)?
- [x] mouse scroll scrollbar scrolling always scrolls up (general sdl2 backend bug)
- [x] readded textarea editing support
- [x] starting dialog closes all other windows and subwindows
- [x] pst smoothly scrolls between dialog states — doesn't instantly snap to top
- [x] mousewheel does not change to scroll the area that is under the cursor (visible in windows with multiple scrollbars)
- [x] keyboard shortcuts assignment to tooltips is missing for menu window (return to game, inventory …)

New features and fixes:
- basic resolution independence
- Subtitle support for BIK videos (SDL2 only probably).
- Video playback in SDL2 is no longer a lag-fest.
- window dragging support (also improving the PST float window)
- tooltip artifacts are cured (very noticeable on the game options). Probably some other tooltip bugs like wrapping too.
- improved debug console
- movement animations are smoother

[A comprehensive writeup](https://github.com/gemrb/gemrb/wiki/Subviews---Origins-and-Summary-of-Changes)

**Comment by MarcelHB** [comment]
Any idea how to proceed? There are some major things to address that I don't find any issue for in either tracker, mostly stuff related to larger screen resolutions, but also map exploration, mysterious fading animations, reloading in PST with widescreen looks severely broken, IWD2 shows no container inventory, ... lots of minor stuff.

IWD was the only game that I couldn't find anything looking too bad within a few minutes of checking.

**Comment by lynxlynxlynx** [comment]
I've now sent an email to Brad to bring him up to speed and talk about the plans.

In general I think we'll just have to go through bit by bit (I don't think all the guiscripts have been ported yet), but we now also have a bigger array of static analysis tools at our disposal, so we can also use that for general stabilisation. 

And if I remember correctly, for some reason WF_BORDERLESS is set on most main game windows. Might be related to chiv's idea for centering the window on higher resolutions, but not sure, since it's doable in the scripts themselves.

**Comment by lynxlynxlynx** [comment]
So, Brad is busy until the end of the year, but already pushed some of this clean stashes. Part of the particle system update is still missing, but other than that, nobody knows more (for desktop), so it's just about the stuff in the tracker and checking everything. Same plan.

**Comment by bradallred** [comment]
We should start opening issues for anything that doesn't have one yet. It will be helpful to have saves for any rendering issues encountered. I'm sure there are rendering bugs/unfinished work, but I don't know of a good way to know where those issues are other than encountering them :(

I didn't try anything with widescreen mod, but I'm not sure widescreen mod adds anything valuable to this branch (I have never used it). You should be able to play any game at any resolutions without it. Still, feel free to open issues; screenshots would be helpful with such issues.

Unfortunately, it appears that SDL1.2 no longer functions at all on Mac. I tried both master and the subviews branch and neither render anything at all. This is going to make it difficult for me to make any fixes for the SDL 1.2 backend. It looks like SDL 1.2 has a lot of commits since 1.2.15 was released, so maybe I can just build it myself and it will work.

However, it would be nice to know what motivation we have for continuing to maintain an SDL 1.2 backend. Are there platforms we support that don't have SDL 2? I know on master SDL 2 is dog slow, but that should be largely mitigated in this branch. There are also a couple of low hanging fruit optimizations we have issues for that could make it better.

**Comment by lynxlynxlynx** [comment]
I've created another issue template for subviews, so reporting will be faster. Not yet pushed though.

Looks like 1.2.16 is on its way: https://github.com/SDL-mirror/SDL/commit/f91fab2d9d0e055383f06d2a07f5901fe905c83c

Looks like we wouldn't miss much, only Pandora was a confirmed runner and it's abandoned as a hw project: https://wiki.libsdl.org/Installation
But let's not remove it before subviews is fixed up, since it would just create more merge problems.

On master SDL2 has other more annoying problems than speed, but subviews fixes at least two, I think: only being able to scroll up with the mouse wheel and something I forgot already.

**Comment by bradallred** [comment]
Good news about SDL 1.2.16; I see several Mac related fixes there.

I'm not proposing we remove it, rather simply make SDL2 the default and stop supporting it. Leave the source, and if somebody wants to use it for some esoteric platform or what have you they can send us a pull request. Ditto OpenGL (which no longer even builds on this branch).

Assuming that SDL 1.2.16 functions for me I intend to do my best to keep it functional until after the subviews branch is merged.

**Comment by bradallred** [comment]
a couple of other fixed issues for the list (if we care to enumerate everything)

- Subtitle support for BIK videos (SDL2 only probably). were there other subtitle issues fixed too?
- Video playback in SDL2 is no longer a lag-fest (the large ones were unwatchable).
- The viewport scrolls when changing focus now. previously it would jump directly to the actor.
- tooltip artifacts are cured (very noticeable on the game options). Probably some other tooltip bugs like wrapping too.
- Actor portraits blink their outline on hover; hopefully somebody will finish this so hovering the actor does the same.

This isn't a bug or missing feature, but it might be worth noting for people that we are fairly resolution independent now:
![image](https://user-images.githubusercontent.com/1126744/64234315-4b576f80-ceb3-11e9-93d3-3e75b544b970.png)

**Comment by lynxlynxlynx** [comment]
Hovering on actors does pulse, but did you mean vice-versa, it also triggering portrait pulsating when possible?

Added the rest to the first post; there has been too much work for me to go through the log, so these will serve as a basis for the changelog and announcements.

**Comment by bradallred** [comment]
oh yeah, that does work... I had that on my list as broken. Neat; must have forgotten fixing that.

One other combination bug fix and feature:
PST floating window dragging is no longer buggy (IIRC it works great unless there is a regression). I've also gone ahead and exposed the window dragging functionality for any window as long as you have the control key down when you drag the window (the window itself, not one of the controls on the window). This behavior is also achievable via a flag on the window if somebody wants to mod things with GUIScript.

![image](https://user-images.githubusercontent.com/1126744/64306002-8e1a5580-cf4e-11e9-9e72-d53621fb0f15.png)

**Comment by MarcelHB** [comment]
@bradallred I have made it through all non-EE games and focused on UI issues, the list contains 40 items that I haven't found in this tracker yet, including screenshots where helpful. What you want me to do? An issue for each one, one large one with checklist, one per game, ...?

**Comment by bradallred** [comment]
> @bradallred I have digged through all non-EE games and focused on UI issues, the list contains 40 items that I haven't found in this tracker yet, including screenshots where helpful. What you want me to do? An issue for each one, one large one with checklist, one per game, ...?

You can probably dump it here and I can open new issues if need be. I would _hope_ some things on your list have been fixed as we progressed. I don't expect you to go verify, but I do hope you have area and coordinates for any rendering issues that remain.

**Comment by MarcelHB** [comment]
Ok, I'll gonna post one issue with information per comment later today. You can then refer to the items individually.

**Comment by lynxlynxlynx** [comment]
I suggest you just open them as issues. It would save a lot of work and confusion if any discussions start, since github is basically single-threaded.

**Comment by MarcelHB** [comment]
I'll do it like that, I just don't want to flood the tracker without asking first. :slightly_smiling_face:

**Comment by MarcelHB** [comment]
I'm done with my current state of "the list". I propose to assign (or unassign again) myself to whatever I can solve, so @bradallred can check the assignment in order to prevent things from getting done twice.

**Comment by lynxlynxlynx** [comment]
Do you guys think there's any hope we can merge this back into master? I fear we'll just have archive it and rename subviews to master once it's ready.

**Comment by bradallred** [comment]
what do you mean exactly? if we did `git merge subviews` thats exactly what git would end up doing (just slapping the master label on the HEAD of subviews) assuming that subviews contains all of masters commits. nobody would even notice.

**Comment by lynxlynxlynx** [comment]
I'm not sure that's true, but I now checked and the only conflicts for merging back are from the code that hasn't been merged to subviews yet. Cool! An archive branch will be enough, for easier bisections not related to this work.

**Comment by bradallred** [comment]
we need to devise a test plan and probably introduce a tag to tag any issues that are merge blockers. I know we are still a ways out, but its helpful to know ahead of time and it might spur somebody to start testing and finding additional issues; the sooner the better.

**Comment by lynxlynxlynx** [comment]
I'm not sure a separate test plan would help. We already have a backlog of issues and testing goals are clear: all the GUIs should work. Once we have them and drawing under our belt, then we can do the final testing that will include prolonged gameplay.

Each time I fire it up, new things are still apparent, so I think it's better to start thinking about this once the supply starts running low.

**Comment by bradallred** [comment]
I was just thinking of trying to compile a list of things to extensively test the rendering pipelines. There are a lot of combinations needing to be tested across 2.5. Even if that list is just a list of the closed rendering issues encountered that would be great for at least being reasonably confidant about minimal regression.

**Comment by lynxlynxlynx** [comment]
What do you mean by 2.5? Anyway, I think you underestimate the sometimes excessive attention to detail our playtesters have. I don't think we need anything extra besides what we have in release criteria already, save perhaps testing global tint (dream areas, timestop) if the full playthrough won't be of bg2.

**Comment by bradallred** [comment]
I meant SDL1.2, SDL2, and SDL2+GL

Its not underestimating _them_ it is knowing things slip past me all the time :)

If you feel confidant, that is good enough for me. However, now that SDL2 uses a separate backend it is actually important for testers to test at least SDL1.2 and one of the SDL2 flavors. Testers tend to just favor whatever the default is so we just need to communicate to them I suppose.

**Comment by bradallred** [comment]
I've added a [wiki page](https://github.com/gemrb/gemrb/wiki/Subviews---Origins-and-Summary-of-Changes) to try and better organize what this branch changes.

**Comment by lynxlynxlynx** [comment]
Very nice, outlines the extent and gravity.

**Comment by MarcelHB** [comment]
As a general info: There is still an amount of script errors that bites me randomly across games. So far, I had trouble to reliably reproduce them. I see two options:

1. I don't open issues until I can describe how to reproduce.
2. I create issues and with some context and thy Python errors, and we think of what it may be.

**Comment by bradallred** [comment]
If you have a python error and trace then that is certainly good enough to open an issue.

I will point out tho: if you forget to do a clean script install after coming from master you will get all sorts of nasty surprise errors.

**Comment by MarcelHB** [comment]
Nah, subviews is in a different directory simply for to save all the time it takes to switch. :wink: Well then, I will sample these errors then.

**Comment by lynxlynxlynx** [comment]
Hey, a positive surprise today. The pathfinder upgrade is 99% done and then we saw that walking stutters a bit if you watch carefully. Well, turns out it's the same on master, but already fixed in subviews. :)

**Comment by bradallred** [comment]
it would be interesting to hear why that would be if there is any speculation about it. nothing comes to my mind immediately.

**Comment by lynxlynxlynx** [comment]
Yeah, I was surprised as well. I suspect it's a side effect of the Draw/DrawActorSprite cleanups.

**Comment by bradallred** [comment]
yes, that would make sense if the pathing updates are happening in `Actor::Draw`, but then we likely have a problem with pathing updates _not_ happening for offscreen actors now. Unless you addressed this in the merge and put it somewhere else.

**Comment by lynxlynxlynx** [comment]
No, that's the odd bit — the pathfinder is elsewhere and the changes are well localised. Let's just treat it like the serendipitous thing that it is.

**Comment by lynxlynxlynx** [comment]
We're in the final stretch! 🔥 
The GUIs have been thoroughly tested and besides a few scratches still on the tracker, should be pretty much fine.

Considering also the branche's bugtracker and the fact most items there are enhancements that can wait, we're at only 15 reported issues in total. :)

**Comment by lynxlynxlynx** [comment]
Ran the latest numbers from todo-grapher to get a general feel for maturity:

> master:
> 2020-12-06 14.769611 10.495311 9.268037
> subviews:
> 2020-12-05 16.231277 12.524922 8.392550

The number of TODOs and FIXMEs is still higher, but the number of HACKs is lower.

**Comment by bradallred** [comment]
sure, tho probably many of the `TODO`s added are not strictly needed. there are several for supporting horizontal scrollbars for example. Lots of stuff in the video drivers for combinations that aren't used either, like supporting `BLIT_HALFTRANS` and `BLIT_COLOR_MOD` for drawing primitives. Pretty sure there are several that will get resolved with #232 as well

**Comment by lynxlynxlynx** [comment]
88 text files were added and 27 got removed, so the adjusted delta is surely bigger. But that also means the keyword density is worse. ;)

The counts for the guiscripts and its plugin actually match. Same for SDLVideo, but 61 more in core (108 are new/changed, 47 were fixed/changed). This is the list to look at: http://paste.debian.net/1175906/

**Comment by bradallred** [comment]
it's pretty clear from that listing that many of these were added relating to existing code so I still feel pretty good about the actual number being reduced.

**Comment by lynxlynxlynx** [comment]
Before we get too distracted ... how much is missing from the list of tagged issues? Things like important stashes, further rewrites and undocumented crucial todos.
https://github.com/gemrb/gemrb/issues?q=is%3Aissue+is%3Aopen+label%3Asubviews

**Comment by bradallred** [comment]
I have some drag+drop event stuff that still needs to go in. It also addresses the cursor issues assigned to me.

**Comment by lynxlynxlynx** [comment]
So besides that one known crash, drawing changes are done?

**Comment by bradallred** [comment]
I know of nothing that can't wait till merge and be submitted as a PR. That said, I've not exactly done extensive testing so who knows what will shake out of hitting it with a hammer.

**Comment by lynxlynxlynx** [comment]
Cool, I think I'll do some more deep game testing then. First need to inspect those pesky overlays.

**Comment by lynxlynxlynx** [comment]
![image](https://user-images.githubusercontent.com/121515/114784972-425f9d00-9d7c-11eb-867b-a973d9ea8542.png)
While progress may feel asymptotic, we're really getting close. :) Is there something in particular blocking that drag work?

**Comment by bradallred** [comment]
some bug I dont recall the specifics of. just a pain to debug drag events for obvious reasons.

**Comment by lynxlynxlynx** [comment]
Well, it's now the last thing on the list. :)

**Comment by bradallred** [comment]
I'm looking at [the diff](https://github.com/gemrb/gemrb/compare/subviews) for suspicious differences (found and fixed some) and I notice that lost of the demo data is different. is that expected?

**Comment by bradallred** [comment]
it also looks like we removed the GUIScript docs?

**Comment by lynxlynxlynx** [comment]
Demo: yes, everything in subviews should have precedence. Dialog.tlk is a binary blob, so merges weren't simple, so I just continued the work on subviews.

The GUIScript docs are automatically generated from the ones in the source, so there's only one version to maintain.

**Comment by bradallred** [comment]
so we should remove the txt docs added for the `Views`?

**Comment by lynxlynxlynx** [comment]
Yeah or inline them as the rest, if they don't exist as PyDoc strings yet.

**Comment by lynxlynxlynx** [comment]
Also went through the diff, just looking at the less common stuff. All seems fine. :)

And wow, the first commit was in January 2015. Time flies!

**Comment by lynxlynxlynx** [comment]
Is there something I can do to help move the drag work along?

**Comment by bradallred** [comment]
I haven't had the time or willpower to get around to fixing the inventory/drag+drop, but its not functionality that exists in master so it probably shouldn't stop up from proceeding. I've pushed the necessary changes to close the last remaining issues.

**Comment by lynxlynxlynx** [comment]
That works for me. :) 

I also gave it some thought and decided to do two releases, since so many changes accrued. One final 0.8.8 without full-scale testing, then we merge, then proper testing starts for 0.9.0. There's at least one regression I'd like to work around first, but other than that, the whole trio is doable within a week, unless some nasty blocker pops up. 🥳

**Comment by bradallred** [comment]
sounds like a sensible choice

**Comment by lynxlynxlynx** [comment]
So new milestones:
- 0.9.0 for any subviews stuff before the release, currently containing just this issue and the subviews-recheck ones
- 0.9.1 is the next regular one, currently with postponed stuff from before

**Comment by lynxlynxlynx** [comment]
All is set, 0.8.8 is out and about. Do you want to do the honours (pull)? 😃

**Comment by bradallred** [comment]
with pleasure sir!

**Comment by lynxlynxlynx** [comment]
The convergence is done, the branch is over! 🎉 🥳 

Great work Brad!

I'd send cookies, but I doubt they'd make it over the pond intact. 🥏
